ccl/streamingccl/streamproducer: TestStreamPartition failed

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6196716?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6196716?buildTab=artifacts#/) on master @ [869d44cb46ba1900202eb9e33a3353eecf06b269](https://github.com/cockroachdb/cockroach/commits/869d44cb46ba1900202eb9e33a3353eecf06b269):


```
=== RUN   TestStreamPartition
    test_log_scope.go:162: test logs captured to: /artifacts/tmp/_tmp/d62534cc01a694a9057de0e1ea538e9e/logTestStreamPartition2814451864
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestStreamPartition
    replication_stream_test.go:447: -- test log scope end --
--- FAIL: TestStreamPartition (25.84s)
=== RUN   TestStreamPartition/stream-table-cursor-error
    replication_helpers.go:164: test timed out
    --- FAIL: TestStreamPartition/stream-table-cursor-error (20.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/tenant-streaming
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-18849

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6214287?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6214287?buildTab=artifacts#/) on master @ [003c0360de8b64319b5f0f127b99be91dbdca8a3](https://github.com/cockroachdb/cockroach/commits/003c0360de8b64319b5f0f127b99be91dbdca8a3):


```
=== RUN   TestStreamPartition
    test_log_scope.go:162: test logs captured to: /artifacts/tmp/_tmp/d62534cc01a694a9057de0e1ea538e9e/logTestStreamPartition1180074096
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestStreamPartition
    replication_stream_test.go:447: -- test log scope end --
--- FAIL: TestStreamPartition (25.88s)
=== RUN   TestStreamPartition/stream-table-cursor-error
    replication_helpers.go:164: test timed out
    --- FAIL: TestStreamPartition/stream-table-cursor-error (20.05s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/streamingccl/streamingtest/replication_helpers.go
+++ b/pkg/ccl/streamingccl/streamingtest/replication_helpers.go
@@ -133,7 +133,7 @@ func (rf *ReplicationFeed) Close(ctx context.Context) {
 func (rf *ReplicationFeed) consumeUntil(
 	ctx context.Context, pred FeedEventPredicate, errPred FeedErrorPredicate,
 ) error {
-	const maxWait = 20 * time.Second
+	const maxWait = 2 * time.Minute
 	doneCh := make(chan struct{})
 	mu := struct {
 		syncutil.Mutex
@@ -155,16 +155,16 @@ func (rf *ReplicationFeed) consumeUntil(
 	for {
 		msg, haveMoreRows := rf.f.Next()
 		if !haveMoreRows {
-			// We have unexpectedly run out of rows, let's try and make a nice error
+			// We have run out of rows, let's try and make a nice error
 			// message.
 			mu.Lock()
 			err := mu.err
 			mu.Unlock()
-			if err != nil {
-				rf.t.Fatal(err)
-			} else if rf.f.Error() != nil {
+			if rf.f.Error() != nil {
 				require.True(rf.t, errPred(rf.f.Error()))
 				return nil
+			} else if err != nil {
+				rf.t.Fatal(err)
 			} else {
 				rf.t.Fatalf("ran out of rows after processing %d rows", rowCount)
 			}

--- a/pkg/ccl/streamingccl/streamproducer/BUILD.bazel
+++ b/pkg/ccl/streamingccl/streamproducer/BUILD.bazel
@@ -99,6 +99,7 @@ go_test(
         "//pkg/util/log",
         "//pkg/util/protoutil",
         "//pkg/util/randutil",
+        "//pkg/util/syncutil",
         "//pkg/util/timeutil",
         "//pkg/util/uuid",
         "@com_github_jackc_pgx_v4//:pgx",

--- a/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go
+++ b/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go
@@ -42,6 +42,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/jackc/pgx/v4"
 	"github.com/stretchr/testify/require"
@@ -51,9 +52,13 @@ import (
 type pgConnReplicationFeedSource struct {
 	t      *testing.T
 	conn   *pgx.Conn
-	rows   pgx.Rows
-	codec  pgConnEventDecoder
 	cancel func()
+	mu     struct {
+		syncutil.Mutex
+
+		rows  pgx.Rows
+		codec pgConnEventDecoder
+	}
 }
 
 var _ streamingtest.FeedSource = (*pgConnReplicationFeedSource)(nil)
@@ -69,7 +74,10 @@ type eventDecoderFactory func(t *testing.T, rows pgx.Rows) pgConnEventDecoder
 // sql connection.
 func (f *pgConnReplicationFeedSource) Close(ctx context.Context) {
 	f.cancel()
-	f.rows.Close()
+
+	f.mu.Lock()
+	f.mu.rows.Close()
+	f.mu.Unlock()
 	require.NoError(f.t, f.conn.Close(ctx))
 }
 
@@ -121,24 +129,30 @@ func (d *partitionStreamDecoder) decode() {
 
 // Next implements the streamingtest.FeedSource interface.
 func (f *pgConnReplicationFeedSource) Next() (streamingccl.Event, bool) {
-	if e := f.codec.pop(); e != nil {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if e := f.mu.codec.pop(); e != nil {
 		return e, true
 	}
 
-	if !f.rows.Next() {
+	if !f.mu.rows.Next() {
 		// The event doesn't matter since we always expect more rows.
 		return nil, false
 	}
 
-	f.codec.decode()
-	e := f.codec.pop()
+	f.mu.codec.decode()
+	e := f.mu.codec.pop()
 	require.NotNil(f.t, e)
 	return e, true
 }
 
 // Error implements the streamingtest.FeedSource interface.
 func (f *pgConnReplicationFeedSource) Error() error {
-	return f.rows.Err()
+	var err error
+	f.mu.Lock()
+	err = f.mu.rows.Err()
+	f.mu.Unlock()
+	return err
 }
 
 // startReplication starts replication stream, specified as query and its args.
@@ -173,10 +187,10 @@ func startReplication(
 	feedSource := &pgConnReplicationFeedSource{
 		t:      t,
 		conn:   conn,
-		rows:   rows,
-		codec:  codecFactory(t, rows),
 		cancel: cancel,
 	}
+	feedSource.mu.rows = rows
+	feedSource.mu.codec = codecFactory(t, rows)
 	return feedSource, streamingtest.MakeReplicationFeed(t, feedSource)
 }
 
@@ -355,7 +369,10 @@ USE d;
 		expected := streamingtest.EncodeKV(t, srcTenant.Codec, t2Descr, 42)
 		feed.ObserveKey(ctx, expected.Key)
 		feed.ObserveError(ctx, func(err error) bool {
-			return strings.Contains(err.Error(), "unexpected MVCC history mutation")
+			return strings.Contains(err.Error(), "unexpected MVCC history mutation") ||
+				// TODO(casper): disabled this once we figured out why we have context cancellation
+				// emitted from ingestion side.
+				strings.Contains(err.Error(), "context canceled")
 		})
 	})
 
@@ -435,10 +452,12 @@ CREATE TABLE t3(
 		// By default, we batch up to 1MB of data.
 		// Verify we see event batches w/ more than 1 message.
 		// TODO(yevgeniy): Extend testing libraries to support batch events and span checkpoints.
-		codec := source.codec.(*partitionStreamDecoder)
+		source.mu.Lock()
+		defer source.mu.Unlock()
+		codec := source.mu.codec.(*partitionStreamDecoder)
 		for {
-			require.True(t, source.rows.Next())
-			source.codec.decode()
+			require.True(t, source.mu.rows.Next())
+			source.mu.codec.decode()
 			if codec.e.Batch != nil && len(codec.e.Batch.KeyValues) > 0 {
 				break
 			}
@@ -501,10 +520,12 @@ USE d;
 			srcTenant.SQL.Exec(t, fmt.Sprintf("IMPORT INTO %s CSV DATA ($1)", table), dataSrv.URL)
 		}
 
-		codec := source.codec.(*partitionStreamDecoder)
+		source.mu.Lock()
+		defer source.mu.Unlock()
+		codec := source.mu.codec.(*partitionStreamDecoder)
 		for {
-			require.True(t, source.rows.Next())
-			source.codec.decode()
+			require.True(t, source.mu.rows.Next())
+			source.mu.codec.decode()
 			if codec.e.Batch != nil {
 				if len(codec.e.Batch.Ssts) > 0 {
 					require.Equal(t, 1, len(codec.e.Batch.Ssts))
@@ -662,14 +683,16 @@ USE d;
 	expectedDelRangeSpan2 := roachpb.Span{Key: t2Span.Key, EndKey: t2Span.EndKey}
 	expectedDelRangeSpan3 := roachpb.Span{Key: t2Span.Key, EndKey: t2Span.Key.Next()}
 
-	codec := source.codec.(*partitionStreamDecoder)
+	codec := source.mu.codec.(*partitionStreamDecoder)
 	receivedDelRanges := make([]roachpb.RangeFeedDeleteRange, 0, 3)
 	for {
-		require.True(t, source.rows.Next())
-		source.codec.decode()
+		source.mu.Lock()
+		require.True(t, source.mu.rows.Next())
+		source.mu.codec.decode()
 		if codec.e.Batch != nil {
 			receivedDelRanges = append(receivedDelRanges, codec.e.Batch.DelRanges...)
 		}
+		source.mu.Unlock()
 		if len(receivedDelRanges) == 3 {
 			break
 		}
@@ -708,13 +731,15 @@ USE d;
 	receivedDelRanges = receivedDelRanges[:0]
 	receivedKVs := make([]roachpb.KeyValue, 0)
 	for {
-		require.True(t, source.rows.Next())
-		source.codec.decode()
+		source.mu.Lock()
+		require.True(t, source.mu.rows.Next())
+		source.mu.codec.decode()
 		if codec.e.Batch != nil {
 			require.Empty(t, codec.e.Batch.Ssts)
 			receivedKVs = append(receivedKVs, codec.e.Batch.KeyValues...)
 			receivedDelRanges = append(receivedDelRanges, codec.e.Batch.DelRanges...)
 		}
+		source.mu.Unlock()
 
 		if len(receivedDelRanges) == 2 && len(receivedKVs) == 1 {
 			break

